arctic/pluggable/generic_version_store.py

```diff
@@ -0,0 +1,436 @@
+            Override the default behavior for allowing reads from secondary members of a cluster:
+            `None` : use the settings from the top-level `Arctic` object used to query this version store.
+            `True` : allow reads from secondary members
+            `False` : only allow reads from primary members
+        """
+        _version = self._backing_store.read_version(self.library_name, symbol, as_of)
+        return VersionedItem(symbol=symbol, library=self.library_name, version=_version['version'],
+                             metadata=_version.pop('metadata', None), data=None)
+
+    def _insert_version(self, version):
+        try:
+            self._backing_store.write_version(self.library_name, version['symbol'], version)
+        except DuplicateKeyError as err:
+            logger.exception(err)
+            raise OperationFailure("A version with the same _id exists, force a clean retry")
+
+
+    def append(self, symbol, data, metadata=None, prune_previous_version=True, upsert=True, **kwargs):
+        """
+        Append 'data' under the specified 'symbol' name to this library.
+        The exact meaning of 'append' is left up to the underlying store implementation.
+
+        Parameters
+        ----------
+        symbol : `str`
+            symbol name for the item
+        data :
+            to be persisted
+        metadata : `dict`
+            an optional dictionary of metadata to persist along with the symbol.
+        prune_previous_version : `bool`
+            Removes previous (non-snapshotted) versions from the database.
+            Default: True
+        upsert : `bool`
+            Write 'data' if no previous version exists.
+        """
+        raise NotImplementedError()
+
+    def write(self, symbol, data, metadata=None, prune_previous_version=True, **kwargs):
+        """
+        Write 'data' under the specified 'symbol' name to this library.
+
+        Parameters
+        ----------
+        symbol : `str`
+            symbol name for the item
+        data :
+            to be persisted
+        metadata : `dict`
+            an optional dictionary of metadata to persist along with the symbol.
+            Default: None
+        prune_previous_version : `bool`
+            Removes previous (non-snapshotted) versions from the database.
+            Default: True
+        kwargs :
+            passed through to the write handler
+
+        Returns
+        -------
+        VersionedItem named tuple containing the metadata and version number
+        of the written symbol in the store.
+        """
+        _id = bson.ObjectId()
+        version = {'_id': _id, 'symbol': symbol, 'metadata': metadata}
+
+        previous_version = self._backing_store.read_version(self.library_name, symbol)
+
+        handler = self._write_handler(version, symbol, data, **kwargs)
+        handler.write(self._backing_store, self.library_name, version, symbol, data, previous_version, **kwargs)
```

Comment:
I thought about this approach. One worry is that it becomes very complicated with an explosion of possible interactions between the different parts which may or may not make sense e.g. mongodb metadata with parquet serialisation with 2Mb chunking on a S3 backend etc. etc.

Each library would then have to be configured with a particular config to work correctly. I think it might be an abstraction too far.